llvm/lib/Transforms/Vectorize/VPlan.h

```diff
@@ -3716,6 +3759,9 @@ class VPlan {
   /// Return the live-in VPValue for \p V, if there is one or nullptr otherwise.
   VPValue *getLiveIn(Value *V) const { return Value2VPValue.lookup(V); }
 
+  /// Return the list of live-in VPValues available in the VPlan.
+  ArrayRef<VPValue *> getLiveIns() const { return VPLiveIns; }
```

Comment:
OK, I still feel this could lead to confusion/problems though. If I understand correctly, `VPLiveIns` is essentially a superset of `Value2VPValue`? For example, any VPValue contained in the map `Value2VPValue` *should* also exist in `VPLiveIns`? Unless live-ins get deleted from `VPLiveIns` or re-added?

Essentially it sounds like `getLiveIn` should really be called something like `findLiveInForValue` or something like that. And we might want to assert somewhere that the values found in the `Value2VPValue` map also exist in `VPLiveIns`. I'm just a bit worried that we have two ways of recording the same state and there might not be anything to defend them being kept in sync?

Reply:
Thanks, I added an assert to `getLiveIns()` to make sure all VPValues also in `Value2VPValue` are included in `VPLiveIns` . At the moment that's guaranteed, as `getOrAddLiveIn` is the only place that adds to  `Value2VPValue` and `VPLiveIns`. Live-ins are only destroyed when the VPlan is destroyed, so at the moment nothing should remove from either `VPLiveIns` or `Value2VPValue` .

I left the name of `getLiveIn` unchanged for now here as that should probably done separately, although the `ForValue` part is implied by the argument type. WDYT?

Reply:
@david-arm WDYT about the latest naming? I'd be planning to land this soon and potentially iterate on the way live-ins are stored separately if needed.